### ISSUE TITLE
Long Shot: Linking two browser windows

### ISSUE THREAD

**Issue by PeterVdBoogaard** [angular/router#183]
It's a bit exotic, and i have no idea how to do it or if it should even be part of the routing..
But.
I'm working on a multi monitor setup and really miss that i can't move parts of a web app outside the viewport of the browser to another monitor. So If i could have 2 'linked' browser windows that are both part of the same app routing that would be sweet. Maybe for each route have the possibility to tell it to open in the second browser viewport.

Haven't really thought it through, but i do know floatable windows i can put away on a second monitor is one of the main thing i missing going from 'native desktop' to webapps.

**Comment by SanderElias** [comment]
@PeterVdBoogaard. Nice idea. I don't think this is something that belongs in a router. If you design your routes well, you can right-click and 'open in a new tab/window' to split out part of the app. 
You can put some session/user stiff in localStorage to have it available to both. localStorage even fires events if the 'other' window changes things. This might be just enough for your system to work.

Another solution is to use websocket's to communicate between clients. The added benefit is that this would even work on different system (think opening a part of your app on a tablet/phone)

The nicest solution would be to use some P2P solution (based on [webrtc](https://developer.mozilla.org/en-US/docs/Web/Guide/API/WebRTC), however, last time I experimented with this stuff it was highly fragile. (that is a few weeks ago), and the browser support is rather shaky still. 
However when this is sorted out, it's the best solution for this, you can then communicate between browsers (even on different systems) just te way you can now communicate with a web-worker.

**Comment by hsablonniere** [comment]
Hi there,

I've done a lot of experimentation regarding cross tab/window communications and multi-screen Web experiences. Like @SanderElias, I'm not sure this belongs to the router but I'm really interested to experiment @PeterVdBoogaard use cases...

I've done a talk about this in french at Paris-Web : http://www.paris-web.fr/2014/conferences/les-recettes-du-web-multi-ecran.php
I'll do a shorter one in english at ScotlandJS : http://scotlandjs.com/speakers/hubert/

**Comment by btford** [comment]
Interesting. I don't think this is common enough of an issue to warrant  spending much of my own time on, but I want to leave this issue open for discussion. If there's enough interest and a clear path forward, we can start figuring out how support for multiple browser windows would look at a router level.

**Comment by PeterVdBoogaard** [comment]
@SanderElias I agree the router might not be the best place, but i don't think there really is a place in the framework atm to put this. 
Using the 'open in new browser' isn't user friendly enough imho. 
Your suggestions make sense, i've done some playing around with signalR (on an MS stack here) , it could be used to solve the communication as you said. The localStorage i would try to use for window state persistence. Like when i open Visual Studio all the floatables are in the same position i left them, with the same 'subviews'.

@btford Glad you think it's something worth considering. I think this is a typical case of build it and they will come. It would need more than just the router, but the router will be an integral part of it i imagine.

The hardest parts of this usecase i think will be to bring all the linked browser windows to the front when one gains focus. Maybe @hsablonniere has figured something out for this in his experimentations.

**Comment by SanderElias** [comment]
@PeterVdBoogaard 
>The hardest parts of this usecase i think will be to bring all the linked browser windows to the front when one gains focus.

This is indeed a problem, Browsers have no control over their size and/or position. The thing that comes the closest is opening a new "window" where you can address some of the settings, but only once. I'm not even sure if this is possible from a chrome-app (or whatever your browser you fancy)

It can be addressed by a separate (node/native) app that can pick up on rearrangement commands. But that kind of defeats the goal. If you go that way, you can build an [nw.js app](https://github.com/nwjs/nw.js/) If you do that, you don't need anything from angular to get this going.

In a web-app setting the best the router can do, is offer an option to open a target in a new window. (indeed remove the burden of a right-click!) Add some ID to a child window, and let the 'communication' service take care of the rest.
It still is a neat idea, I might build something to play with it. A kind of master-detail view, where the detail is in another window.

**Comment by hsablonniere** [comment]
Glad to hear that it interests @btford. I'll have some time to experiment again in april. I'll try approaches that directly involves routes...

@SanderElias @PeterVdBoogaard I confirm that you can play with popup position and size.

The first problem is to detect the user has more than one monitor. I didn't find a standard way that is cross browser. Flash could be an option :-(

The second problem is that some browsers like Chrome doesn't allow to place a window on another monitor space programmatically.
https://code.google.com/p/chromium/issues/detail?id=137681

I published an experiment I worked on 4 months ago to play with size and position : https://hsablonniere.github.io/window-popup/

**Comment by HipsterZipster** [comment]
+1 Very Interested to see if anything comes out of this without falling back onto Native Runtimes or Flash

**Comment by hsablonniere** [comment]
My talk on multi-screen and multi-tab Web techniques is out ;-)
https://www.youtube.com/watch?v=-o8B4TE18gI

If anyone tries some of theses techniques hidden behind a router API that would be awesome. I want to try that but I need to find some time...

**Comment by shameeryou** [comment]
I'm facing a similar issue that @PeterVdBoogaard faced.  I need three windows in my app and I want to control two other windows using the main window. In other two windows, there will be only a video element that syncs with main window video. I need to control video playback of two other windows from the main window. Any hope?

**Comment by pyummm** [comment]
and any solutions?
